lib/puppet/pops/types/types.rb

```diff
@@ -1307,8 +1524,9 @@ class PVariantType < PAnyType
 
       attr_reader :types
 
+      # @param types [Array[PAnyType]] the variants
       def initialize(types)
-        @types = types.freeze
+        @types = types.uniq.freeze
```

Comment:
ok, it is fine to have it here, but it should really be in the others since it seems it would waste cycles first creating types for all of the integers, and then they are made unique. In your example case it could even be optimized into Integer[0,2]. We should perhaps use a Set when creating the set of types for Variant (not sure how wan avoid creating the type to add though, so we still end up with garbage). --- i.e. leave it like you have it in this PR. May want to optimize some of the inferences later.